pyomo/contrib/fbbt/fbbt.py

```diff
@@ -1084,7 +970,77 @@ def _check_and_reset_bounds(var, lb, ub):
     return lb, ub
 
 
-class _FBBTVisitorLeafToRoot(ExpressionValueVisitor):
+def _before_constant(visitor, child):
+    if child in visitor.bnds_dict:
+        pass
+    else:
+        visitor.bnds_dict[child] = (child, child)
+    return False, None
+
+
+def _before_var(visitor, child):
+    if child in visitor.bnds_dict:
+        return False, None
+    elif child.is_fixed() and not visitor.ignore_fixed:
+        lb = value(child.value)
+        ub = lb
+    else:
+        lb = child.lb
+        ub = child.ub
+        if lb is None:
+            lb = -interval.inf
+        if ub is None:
+            ub = interval.inf
+        if lb - visitor.feasibility_tol > ub:
+            raise InfeasibleConstraintException(
+                'Variable has a lower bound that is larger than its '
+                'upper bound: {0}'.format(str(child))
+            )
+    visitor.bnds_dict[child] = (lb, ub)
+    return False, None
+
+
+def _before_NPV(visitor, child):
+    if child in visitor.bnds_dict:
+        return False, None
+    val = value(child)
+    visitor.bnds_dict[child] = (val, val)
+    return False, None
+
+
+def _before_other(visitor, child):
+    return True, None
+
+
+def _before_external_function(visitor, child):
+    # TODO: provide some mechanism for users to provide interval
+    # arithmetic callback functions for general external
+    # functions
+    visitor.bnds_dict[child] = (-interval.inf, interval.inf)
+    return False, None
+
+
+def _register_new_before_child_handler(visitor, child):
+    handlers = _before_child_handlers
+    child_type = child.__class__
+    if child.is_variable_type():
+        handlers[child_type] = _before_var
+    elif not child.is_potentially_variable():
+        handlers[child_type] = _before_NPV
+    else:
+        handlers[child_type] = _before_other
```

Comment:
I feel like if we hit this, then we should warn about an unknown / unhandled expression type?